providers/vultr/vultr.go

```diff
@@ -0,0 +1,342 @@
+}
+
+// GetNameservers gets the Vultr nameservers for a domain
+func (api *VultrApi) GetNameservers(domain string) ([]*models.Nameserver, error) {
+	return models.StringsToNameservers(defaultNS), nil
+}
+
+// EnsureDomainExists adds a domain to the Vutr DNS service if it does not exist
+func (api *VultrApi) EnsureDomainExists(domain string) error {
+	ok, err := api.isDomainInAccount(domain)
+	if err != nil {
+		return err
+	}
+
+	if !ok {
+		// Vultr requires an initial IP, use a dummy one
+		err := api.client.CreateDNSDomain(domain, "127.0.0.1")
+		if err != nil {
+			return err
+		}
+
+		ok, err := api.isDomainInAccount(domain)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			return fmt.Errorf("Unexpected error adding domain %s to Vultr account", domain)
+		}
+	}
+
+	return nil
+}
+
+func (api *VultrApi) isDomainInAccount(domain string) (bool, error) {
+	domains, err := api.client.GetDNSDomains()
+	if err != nil {
+		return false, err
+	}
+
+	var vd *vultr.DNSDomain
+	for _, d := range domains {
+		if d.Domain == domain {
+			vd = &d
+		}
+	}
+
+	if vd == nil {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// toRecordConfig converts a Vultr DNSRecord to a RecordConfig
+func toRecordConfig(dc *models.DomainConfig, r *vultr.DNSRecord) (*models.RecordConfig, error) {
+	// Turns r.Name into a FQDN
+	// Vultr uses "" as the apex domain, instead of "@", and this handles it fine.
+	name := dnsutil.AddOrigin(r.Name, dc.Name)
+
+	data := r.Data
+	// Make target into a FQDN if it is a CNAME, NS, MX, or SRV #rtype_variations
+	if r.Type == "CNAME" || r.Type == "NS" || r.Type == "MX" {
+		if !strings.HasSuffix(data, ".") {
+			data = data + "."
+		}
+		data = dnsutil.AddOrigin(data, dc.Name)
+	}
+	// Remove quotes if it is a TXT #rtype_variations
+	if r.Type == "TXT" {
+		if !strings.HasPrefix(data, `"`) || !strings.HasSuffix(data, `"`) {
+			return nil, errors.New("Unexpected lack of quotes in TXT record from Vultr")
+		}
+		data = data[1 : len(data)-1]
+	}
+
+	rc := &models.RecordConfig{
+		NameFQDN: name,
+		Type:     r.Type,
+		Target:   data,
+		TTL:      uint32(r.TTL),
+		Original: r,
+	}
+
+	// #rtype_variations
+	if r.Type == "MX" {
+		rc.MxPreference = uint16(r.Priority)
+	}
+
+	// #rtype_variations
+	if r.Type == "SRV" {
+		rc.SrvPriority = uint16(r.Priority)
+
+		// Vultr returns in the format "[weight] [port] [target]"
+		splitData := strings.SplitN(rc.Target, " ", 3)
+		if len(splitData) != 3 {
+			return nil, fmt.Errorf("Unexpected data for SRV record returned by Vultr")
+		}
+
+		weight, err := strconv.ParseUint(splitData[0], 10, 16)
+		if err != nil {
+			return nil, err
+		}
+		rc.SrvWeight = uint16(weight)
+
+		port, err := strconv.ParseUint(splitData[1], 10, 16)
+		if err != nil {
+			return nil, err
+		}
+		rc.SrvPort = uint16(port)
+
+		target := splitData[2]
+		if !strings.HasSuffix(target, ".") {
+			target = target + "."
+		}
+		rc.Target = dnsutil.AddOrigin(target, dc.Name)
+	}
+
+	// #rtype_variations
```

Comment:
I don't think we need this #comment everywhere here. Just at the top so this function shows up in grep.

Reply:
Ok, the docs on adding rtypes were not all that clear, I guess I will fix that next.

Reply:
Done